README.md
# Dead Simple Sitemap Generator 0.5

This script will crawl all unique internal links found on a given website. 

## How to Install

Drop this into your `composer.json`
manifest file:

    {
        "require": {
            "gsavastano/dssg": "0.5"
        }
    }

Then run `composer install`.

## Setup

[optional] create a config file as follows

	{
		"protocol": "http",
		"url": "example.com",
		"extension": [
			".htm",
			".html",
			".php",
			"/",
			".aspx",
			".asp"
		],
		"filename": "sitemap.xml",
		"priority": "0.3",
		"frequency": "daily"
	}

you can use the command line to ovveride all parameters except the extensions list

if no config file is provided, the following list of extensions will be scanned:

- .htm
- .html
- .php
- .asp
- .aspx
- /

## Getting Started

Create a simple PHP file

	#!/usr/bin/env php
    <?php
	
	require 'vendor/autoload.php';
	use gsavastano\dssg;
	
	$crawler = new Dssg\Crawl;
	
	//optional if passing params by CLI
	$crawler->loadConfig(__DIR__.'/config.json');
	
	$crawler->startCrawl();
	
then launch it from the command line

	php yourfile.php -uexample.com -ahttps -p0.2 -fdaily -sexample.xml

to check if you set the right values

	php yourfile.php -uexample.com -ahttps -p0.2 -fdaily -sexample.xml -c

to see the options

	php yourfile.php -h

please consider: the script WILL timeout with big/huge websites, if that happens try increasing your memory_limit

	ini_set('memory_limit', '2048M');
	

## Limitations

- Url Validation relies on http code 200, if your ISP has a curtesy page for inexistent domains, the you'll get a false positive
- it is not possible to specify extensions list with CLI
- doesn't always parse correctly all URL parameters (ex: ?#&)
- all URL indexed have the same priority and frequency
- it does not index assets
- it does not index links pointing to external websites
- it does not index subdomains
- basic test coverage

## System Requirements

* PHP 5.3.0+

## License

MIT Public License

CHANGELOG.md
### v. 0.5 - 26/Jul/2016
- relative url to absolute url improved
- bug fixing

### v. 0.4 - 20/Jul/2016
- new Validate class
- psr-4 autoload
- better error handling

### v. 0.3.2 - 19/Jul/2016
- small cosmetic changes

### v. 0.3.1 - 19/Jul/2016
- bug fixing

### v. 0.3 - 19/Jul/2016
- removed external dependencies
- faster url parsing

### v. 0.2 - 19/Jul/2016
- complete refactoring as composer package
- PSR-0 and PSR-2 valid
- more tests
- much lower NPath and Cyclomatic complexity (phpmd)
- full validation of CLI params
- can use config file to store default params
- new option "-c": let you print the configuration on CLI
- bug fixes

### v. 0.1 17/Jul/2016
- first revision

tests/DssgTest.php
<?php
 
require_once __DIR__.'/../src/Dssg.php';
use gsavastano\Dssg\Dssg;
 
class DssgTest extends PHPUnit_Framework_TestCase {
 
    public function testValProtocol()
    {
        $method = new ReflectionMethod('gsavastano\Dssg\Dssg', 'valProtocol');
        $method->setAccessible(TRUE);
 
        $this->assertTrue( $method->invokeArgs(new Dssg, array('http')));
        $this->assertTrue( $method->invokeArgs(new Dssg, array('https')));
        $this->assertFalse( $method->invokeArgs(new Dssg, array('non-http')));
    }

	public function testValFileName()
	{
        $method = new ReflectionMethod('gsavastano\Dssg\Dssg', 'valFileName');
        $method->setAccessible(TRUE);

		$this->assertTrue($method->invokeArgs(new Dssg, array('sitemap.xml')));
		$this->assertTrue($method->invokeArgs(new Dssg, array('site_map_2.xml')));
		$this->assertFalse($method->invokeArgs(new Dssg, array('site_map.it.xml')));
		$this->assertFalse($method->invokeArgs(new Dssg, array('site_map.it')));
	}

	public function testValFrequency()
	{
        $method = new ReflectionMethod('gsavastano\Dssg\Dssg', 'valFrequency');
        $method->setAccessible(TRUE);

		$this->assertTrue($method->invokeArgs(new Dssg, array('always')));
		$this->assertTrue($method->invokeArgs(new Dssg, array('hourly')));
		$this->assertTrue($method->invokeArgs(new Dssg, array('daily')));
		$this->assertTrue($method->invokeArgs(new Dssg, array('weekly')));
		$this->assertTrue($method->invokeArgs(new Dssg, array('monthly')));
		$this->assertTrue($method->invokeArgs(new Dssg, array('yearly')));
		$this->assertTrue($method->invokeArgs(new Dssg, array('never')));
		$this->assertFalse($method->invokeArgs(new Dssg, array('falseAlways')));
		$this->assertFalse($method->invokeArgs(new Dssg, array('falseHourly')));
		$this->assertFalse($method->invokeArgs(new Dssg, array('falseDaily')));
		$this->assertFalse($method->invokeArgs(new Dssg, array('falseWeekly')));
		$this->assertFalse($method->invokeArgs(new Dssg, array('falseMonthly')));
		$this->assertFalse($method->invokeArgs(new Dssg, array('falseYearly')));
		$this->assertFalse($method->invokeArgs(new Dssg, array('falseNever')));
	}

	public function testValPriority()
	{
        $method = new ReflectionMethod('gsavastano\Dssg\Dssg', 'valPriority');
        $method->setAccessible(TRUE);

		$this->assertTrue($method->invokeArgs(new Dssg, array(1)));
		$this->assertTrue($method->invokeArgs(new Dssg, array(0.2)));
		$this->assertFalse($method->invokeArgs(new Dssg, array(0)));
		$this->assertFalse($method->invokeArgs(new Dssg, array(2)));
		$this->assertFalse($method->invokeArgs(new Dssg, array(-1)));
		$this->assertFalse($method->invokeArgs(new Dssg, array('d')));
	}


	public function testValUrl()
	{
		$method = new ReflectionMethod('gsavastano\Dssg\Dssg', 'valUrl');
        $method->setAccessible(TRUE);

		$this->assertTrue($method->invokeArgs(new Dssg, array('http://www.example.com')));
		$this->assertTrue($method->invokeArgs(new Dssg, array('http://example.com')));
		$this->assertTrue($method->invokeArgs(new Dssg, array('https://www.example.com')));
		$this->assertTrue($method->invokeArgs(new Dssg, array('https://example.com')));
		$this->assertFalse($method->invokeArgs(new Dssg, array('httplp://www.example.com')));
		$this->assertFalse($method->invokeArgs(new Dssg, array('http://-&example.com')));
		$this->assertFalse($method->invokeArgs(new Dssg, array('https://www.example')));
		$this->assertFalse($method->invokeArgs(new Dssg, array('http://1.com')));
	}

	public function testRelToAbs()
	{
		$method = new ReflectionMethod('gsavastano\Dssg\Dssg', 'relToAbs');
        $method->setAccessible(TRUE);

        $this->assertEquals('http://www.example.com/rel2abs',$method->invokeArgs(new Dssg, array('/rel2abs','http://www.example.com/')));
		$this->assertEquals('http://www.example.com/../rel2abs',$method->invokeArgs(new Dssg, array('../rel2abs','http://www.example.com/')));
		$this->assertNotEquals('http://www.example.com/rel2abs',$method->invokeArgs(new Dssg, array('http://example.com/rel2abs','http://www.example.com/')));
	}

}

example/crawl-test.php
#!/usr/bin/env php

<?php
//error_reporting(0);
date_default_timezone_set('Europe/Rome');
ini_set('memory_limit', '2048M');

require_once __DIR__.'/../src/Crawl.php';

use gsavastano\dssg;

$crawler = new Dssg\Crawl;

//optional if passing params by CLI
$crawler->loadConfig(__DIR__.'/config.json');

$crawler->startCrawl();

sitemap.php
<?php
	ini_set('memory_limit', '2048M');
	require_once "simple_html_dom.php";

	$defaults = parse_ini_file("sitemap.ini");
	define ('VERSION', $defaults['version']);

	$options = getopt("v::h::c::s::u::f::p::");
	$scanned = array ();

	if(isset($options['h'])) printHelp();
	if(isset($options['v'])) printVersion();

	$config = array(
		'sitemap' 	=> 	valFileName($options['s']) 			? $options['s'] : $defaults['sitemap'],
		'protocol'	=>	valProtocol($options['a'])			? $options['a'] : $defaults['protocol'],
		'url' 		=>	isset($options['u'])				? $options['u'] : $defaults['url'],
		'frequency' => 	valFrequency($options['f']) 		? $options['s'] : $defaults['frequency'],
		'priority' 	=> 	valPriority($options['f']) 			? $options['p'] : $defaults['priority'],
		'extension'	=>	explode(',',$defaults['extension']),
	);

	$config['url'] = filter_var ($config['protocol']."://".$config['url'], FILTER_SANITIZE_URL);

	if(isset($options['c'])) printConfig();

	function valProtocol ($val) {
		$protocols = array('http', 'https');
		return in_array($val, $protocols);
	}

	function valFileName ($val) {
		$pattern = '/^(?!.*\/)(\w|\s|-)+\.xml$/';
		return preg_match($pattern, $val) == 1 ? true : false;
	}
	
	function valFrequency ($val) {
		$frequencies = array('always','hourly','daily','weekly','monthly','yearly','never');
		return in_array($val, $frequencies);
	}
	
	function valPriority ($val) {
		return filter_var( $val, FILTER_VALIDATE_FLOAT	) > 1 || !is_numeric($val) ? false : $val;
	}

	function printHelp() {
		echo '
Dead Simple Sitemap Generator version '.VERSION.'
Giovanni Savastano (gsavastano@gmail.com)
MIT Licence
Use at your own risk :)

Usage: php sitemap.php [options] ...

Option 	Meaning
-v 	Display program version
-h 	Print Help 
-s 	Set Output file name  
-a 	Set Protocol
-u 	Set Target URL 
-f 	Set Frequency
-p 	Set Priority
';
		die();
	}

	function printVersion() {
		echo '
Dead Simple Sitemap Generator version '.VERSION.'
Giovanni Savastano (gsavastano@gmail.com)
MIT Licence
Use at your own risk :)
';
		die();
	}

	function printConfig() {
		global $config;
		foreach ($config as $option => $value) {
			$value = is_array($value) ? implode(',',$value) : $value;
			echo $option." => ".$value."\n";
		}
		die();
	}

	function rel2abs($rel, $base) {
		if(strpos($rel,"//") === 0) {
			return "http:".$rel;
		}
		/* return if  already absolute URL */
		if  (parse_url($rel, PHP_URL_SCHEME) != '') return $rel;
		$first_char = substr ($rel, 0, 1);
		/* queries and  anchors */
		if ($first_char == '#'  || $first_char == '?') return $base.$rel;
		/* parse base URL  and convert to local variables:
		$scheme, $host,  $path */
		extract(parse_url($base));
		/* remove  non-directory element from path */
		$path = preg_replace('#/[^/]*$#',  '', $path);
		/* destroy path if  relative url points to root */
		if ($first_char ==  '/') $path = '';
		/* dirty absolute  URL */
		$abs =  "$host$path/$rel";
		/* replace '//' or  '/./' or '/foo/../' with '/' */
		$re =  array('#(/.?/)#', '#/(?!..)[^/]+/../#');
		for($n=1; $n>0;  $abs=preg_replace($re, '/', $abs, -1, $n)) {}
		/* absolute URL is  ready! */
		return  $scheme.'://'.$abs;
	}

	function getUrl ($url) {
		$agent = "Mozilla/5.0 (compatible;)";

		$ch = curl_init();
		curl_setopt ($ch, CURLOPT_AUTOREFERER, true);
		curl_setopt ($ch, CURLOPT_URL, $url);
		curl_setopt ($ch, CURLOPT_USERAGENT, $agent);
		curl_setopt ($ch, CURLOPT_VERBOSE, 1);
		curl_setopt ($ch, CURLOPT_SSL_VERIFYPEER, FALSE);
		curl_setopt ($ch, CURLOPT_SSL_VERIFYHOST, FALSE);
		curl_setopt ($ch, CURLOPT_HEADER, 0);
		curl_setopt ($ch, CURLOPT_FOLLOWLOCATION, 1);
		curl_setopt ($ch, CURLOPT_RETURNTRANSFER, 1);
		curl_setopt ($ch, CURLOPT_CONNECTTIMEOUT, 5);

		$data = curl_exec($ch);

		curl_close($ch);

		return $data;
	}

	function doScan ($url) {
		global $config, $scanned, $pf;

		$url = filter_var ($url, FILTER_SANITIZE_URL);

		if (!filter_var ($url, FILTER_VALIDATE_URL) || in_array ($url, $scanned)) {
			return;
		}

		array_push ($scanned, $url);
		$html = str_get_html (getUrl ($url));
		if(is_object($html)) {
			$links = $html->find('a');
			unset($html);
		} else 
			return;
		
		if(!(is_array($links) && !empty($links))) return;

		foreach ($links as $val) {
			$next_url = $val->href or "";

			$fragment_split = explode ("#", $next_url);
			$next_url       = $fragment_split[0];

			if ((substr ($next_url, 0, 7) != "http://")  && 
				(substr ($next_url, 0, 8) != "https://") &&
				(substr ($next_url, 0, 6) != "ftp://")   &&
				(substr ($next_url, 0, 7) != "mailto:")) {
				$next_url = @rel2abs ($next_url, $url);
			}

			$next_url = filter_var ($next_url, FILTER_SANITIZE_URL);

			if (substr ($next_url, 0, strlen ($config['url'])) == $config['url']) {
				$ignore = false;

				if (!filter_var ($next_url, FILTER_VALIDATE_URL)) {
					$ignore = true;
				}

				if (in_array ($next_url, $scanned)) {
					$ignore = true;
				}

				if (!$ignore) {
					foreach ($config['extension'] as $ext) {
						if (strpos ($next_url, trim($ext)) > 0) {
							fwrite ($pf, "  <url>\n" .
										 "    <loc>" . htmlentities ($next_url) ."</loc>\n" .
										 "    <changefreq>".$config['frequency']."</changefreq>\n" .
										 "    <priority>".$config['priority']."</priority>\n" .
										 "  </url>\n");
							doScan ($next_url);
						}
					}
				}
			}
		}
	}


	$pf = fopen ($config['sitemap'], "w");
	if (!$pf) {
		echo "Cannot create ".$config['sitemap']."!" . "\n";
		return;
	}

	fwrite ($pf, "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" .
				 "<urlset xmlns=\"http://www.sitemaps.org/schemas/sitemap/0.9\"\n" .
				 "        xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"\n" .
				 "        xsi:schemaLocation=\"http://www.sitemaps.org/schemas/sitemap/0.9\n" .
				 "        http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd\">\n" .
				 "  <url>\n" .
				 "    <loc>" . htmlentities ($config['url']) ."</loc>\n" .
				 "    <changefreq>".$config['frequency']."</changefreq>\n" .
				 "    <priority>".$config['priority']."</priority>\n" .
				 "  </url>\n");

	doScan ($config['url']);
	fwrite ($pf, "</urlset>\n");
	fclose ($pf);

	echo "Done." . "\n";
	echo $config['sitemap']." created." . "\n";
?>

src/Crawl.php
<?php

namespace gsavastano\Dssg;

include_once __DIR__.'/Validate.php';

class Crawl
{
    /**
     * @var array Configuration values
     */
    protected $config = [];

    /**
     * @var array List of Values from CLI
     */
    public $options = [];

    /**
     * @var array List of URLs already scanned
     */
    private $scanned = [];

    /**
     * @var string CLI Arguments
     */
    private $args = 'vhcs::a::u::f::p::';

    /**
     * @var object File Pointer
     */
    private $fileHandle = false;

    /**
     * @var bool Check Script status
     */
    private $ready = false;

    private $validate;

    const _NL = "\n";
    const _VERSION = '0.5';

    public function __construct()
    {
        if (!Validate::isCli()) {
            die('Please use Command Line');
        }
    }

    private function loadStaticConfig($file)
    {
        if (!file_exists($file) || !fopen($file, 'r')) {
            $this->config['error'][] = 'Cannot open configuration file';
        }

        $this->config = json_decode(file_get_contents($file), true);
    }

    private function loadIntercativeConfig()
    {
        $this->options = getopt($this->args);
        if (empty($this->options)) {
            return;
        }
        if (isset($this->options['a'])) {
            $this->config['protocol'] = $this->options['a'];
        }
        if (isset($this->options['u'])) {
            $this->config['url'] = $this->options['u'];
        }
        if (isset($this->options['s'])) {
            $this->config['filename'] = $this->options['s'];
        }
        if (isset($this->options['f'])) {
            $this->config['frequency'] = $this->options['f'];
        }
        if (isset($this->options['p'])) {
            $this->config['priority'] = $this->options['p'];
        }
    }

    public function loadConfig($file = false)
    {
        if ($file) {
            $this->loadStaticConfig($file);
        }
        $this->loadIntercativeConfig();
        $this->config = Validate::validateConfig($this->config);

        if (isset($this->config['error'])) {
            foreach ($this->config['error'] as $error) {
                echo $error.self::_NL;
            }
            die();
        }

        if (isset($this->options['h'])) {
            $this->printHelp();
        }
        if (isset($this->options['v'])) {
            $this->printVersion();
        }
        if (isset($this->options['c'])) {
            $this->printConfig();
        }

        $this->ready = true;

        return;
    }

    private function printConfig($json = false)
    {
        if (!$json) {
            echo '[Crawler Config Variables]'.self::_NL;
            foreach ($this->config as $option => $value) {
                $value = is_array($value) ? implode(', ', $value) : $value;
                echo $option.' => '.$value.self::_NL;
            }
            echo self::_NL;
        } else {
            return json_encode($this->config);
        }
        die();
    }

    public function startCrawl()
    {
        if (!$this->ready) {
            $this->loadConfig();
        }
        $this->createFile();
        $this->doScan($this->config['url']);
        $this->saveAndClose();

        return;
    }

    protected function getUrl($url)
    {
        $agent = 'Mozilla/5.0(compatible;)';
        $curlHandler = curl_init();
        curl_setopt($curlHandler, CURLOPT_AUTOREFERER, true);
        curl_setopt($curlHandler, CURLOPT_URL, $url);
        curl_setopt($curlHandler, CURLOPT_USERAGENT, $agent);
        curl_setopt($curlHandler, CURLOPT_SSL_VERIFYPEER, false);
        curl_setopt($curlHandler, CURLOPT_SSL_VERIFYHOST, false);
        curl_setopt($curlHandler, CURLOPT_HEADER, 0);
        curl_setopt($curlHandler, CURLOPT_FOLLOWLOCATION, 1);
        curl_setopt($curlHandler, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($curlHandler, CURLOPT_CONNECTTIMEOUT, 5);

        $data = curl_exec($curlHandler);

        curl_close($curlHandler);
        return $data;
    }


    function rel2abs($rel, $base)
    {
        $host = $path = '';
        if (parse_url($rel, PHP_URL_SCHEME) != '' || substr($rel, 0, 2) == '//') {
            return $rel;   
        }
        if ($rel[0]=='#' || $rel[0]=='?') {
            return $base.$rel;
        }
        extract(parse_url($base));
        $path = preg_replace('#/[^/]*$#', '', $path);
        if ($rel[0] == '/') {
            $path = '';
        }
        $abs = "$host$path/$rel";
        $re = array('#(/\.?/)#', '#/(?!\.\.)[^/]+/\.\./#');
        for($n=1; $n>0; $abs=preg_replace($re, '/', $abs, -1, $n)) {}
        return $scheme.'://'.$abs;
    }


    protected function clearUrlList(array $list)
    {
        if (empty($list)) {
            return;
        }

        $params = parse_url($this->config['url']);

        $needle = $params['host'];
        $lenght = strlen($this->config['url']);
        if (stripos($needle, 'www') !== false) {
            $needle = substr($needle, 4);
        }
        unset($params);

        //remove duplicates
        $list = array_merge(array_flip(array_flip($list)));
        
        //transform relative links to absolute, remove links to external websites and to subdomains
        foreach ($list as $i => $link) {
            
            $link = $this->rel2abs($link, $this->config['url']);

            //no subdomains or external websites
            if (substr($link, 0, $lenght) !== $this->config['url']) {
                $link = $this->config['url'];
            }
            $list[$i] = filter_var($link, FILTER_SANITIZE_URL);
        }

        //remove all links that have already been scanned.
        foreach ($list as $i => $link) {
            if (in_array($link, $this->scanned)) {
                unset($list[$i]);
            }
        }

        //remove duplicates again
        $list = array_merge(array_flip(array_flip($list)));

        return $list;
    }

    protected function doScan($url)
    {
        if (in_array($url, $this->scanned)) {
            return;
        }
        echo self::_NL.'Scanning: '.$url;
        array_push($this->scanned, $url);

        $dom = new \DOMDocument();
        $internalErrors = libxml_use_internal_errors(true);
        @$dom->loadHTML($this->getUrl($url));
        libxml_use_internal_errors($internalErrors);
        $links = array();
        foreach ($dom->getElementsByTagName('a') as $link) {
            $links[] = $link->getAttribute('href');
        }

        $links = $this->clearUrlList($links);
        
        if (empty($links)) {
            return;
        }
        foreach ($links as $val) {
            $nextUrl = $val or '';

            foreach ($this->config['extension'] as $ext) {
                if (strpos($nextUrl, trim($ext)) > 0) {
                    fwrite(
                        $this->fileHandle,
                        '
                        <url>
                            <loc>'.htmlentities($nextUrl).'</loc>
                            <changefreq>'.$this->config['frequency'].'</changefreq>
                            <priority>'.$this->config['priority'].'</priority>
                        </url>'
                    );
                    $this->doScan($nextUrl);
                }
            }
        }
    }

    private function createFile()
    {
        $this->fileHandle = fopen($this->config['filename'], 'w');
        if (!$this->fileHandle) {
            die('Sitemap File not available'.self::_NL);
        }

        fwrite(
            $this->fileHandle,
            '<?xml version="1.0" encoding="UTF-8"?>
            <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
                    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                    xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9
                    http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd">
            <url>'.self::_NL.'
            <loc>'.htmlentities($this->config['url']).'</loc>
                <changefreq>'.$this->config['frequency'].'</changefreq>
                <priority>'.$this->config['priority'].'</priority>
            </url>'.self::_NL
        );

        return;
    }

    private function saveAndClose()
    {
        if (!$this->fileHandle) {
            die("Can't find sitemap file to close".self::_NL);
        }
        fwrite($this->fileHandle, '</urlset>'.self::_NL);
        fclose($this->fileHandle);

        echo self::_NL.'Done.'.self::_NL.$this->config['filename'].' created.'.self::_NL;
    }

    private function printHelp()
    {
        echo '
Dead Simple Sitemap Generator version '.self::_VERSION.'
Giovanni Savastano(gsavastano@gmail.com)
MIT Licence
Use at your own risk :)

Usage: php sitemap.php [options] ...

Option  Meaning
-v  Display program version
-h  Print Help 
-h  Print Crawler Config
-s  Set Output file name  
-a  Set Procol
-u  Set Target URL 
-f  Set Frequency
-p  Set Priority
';
        die();
    }

    private function printVersion()
    {
        echo '
Dead Simple Sitemap Generator version '.self::_VERSION.'
Giovanni Savastano(gsavastano@gmail.com)
MIT Licence
Use at your own risk :)
';
        die();
    }
}


src/Dssg.php
<?php

namespace gsavastano\Dssg;

class Dssg
{
    /**
     * @var array Configuration values
     */
    protected $config = [];

    /**
     * @var array List of Values from CLI
     */
    public $options = [];

    /**
     * @var array List of URLs already scanned
     */
    private $scanned = [];

    /**
     * @var string CLI Arguments
     */
    private $args = 'vhcs::a::u::f::p::';

    /**
     * @var array Default extensions
     */
    private $defaultExtensions = array('.htm', '.html', '.php', '/', '.aspx', '.asp');

    /**
     * @var object File Pointer
     */
    private $fileHandle = false;

    /**
     * @var bool Check Script status
     */
    private $ready = false;

    const NL = "\n";
    const VERSION = '0.3.2';

    public function __construct()
    {
        if (!$this->isCli()) {
            die('Please use Command Line');
        }
    }

    private function isCli()
    {
        return php_sapi_name() === 'cli';
    }

    private function loadStaticConfig($file)
    {
        if (!file_exists($file) || !fopen($file, 'r')) {
            die('Cannot open '.$file.NL);
        }

        $this->config = json_decode(file_get_contents($file), true);
    }

    private function loadIntercativeConfig()
    {
        $this->options = getopt($this->args);
        if (!empty($this->options)) {
            if (isset($this->options['a'])) {
                $this->config['protocol'] = $this->options['a'];
            }
            if (isset($this->options['u'])) {
                $this->config['url'] = $this->options['u'];
            }
            if (isset($this->options['s'])) {
                $this->config['filename'] = $this->options['s'];
            }
            if (isset($this->options['f'])) {
                $this->config['frequency'] = $this->options['f'];
            }
            if (isset($this->options['p'])) {
                $this->config['priority'] = $this->options['p'];
            }
        }
    }

    public function loadConfig($file = false)
    {
        if ($file) {
            $this->loadStaticConfig($file);
        }
        $this->loadIntercativeConfig();
        $this->validateConfig();

        if (isset($this->options['h'])) {
            $this->printHelp();
        }
        if (isset($this->options['v'])) {
            $this->printVersion();
        }
        if (isset($this->options['c'])) {
            $this->printConfig();
        }

        $this->ready = true;

        return;
    }

    private function validateConfig()
    {
        if (!$this->valFilename($this->config['filename'])) {
            die('Sitemap file name not valid'.self::NL);
        }
        if (!$this->valProtocol($this->config['protocol'])) {
            die('Procol type not valid'.self::NL);
        }

        $completeUrl = $this->config['protocol'].'://'.$this->config['url'];
        $params = parse_url($completeUrl);
        $completeUrl = $params['scheme'].'://'.$params['host'];
        unset($params);

        if (!$this->valUrl($completeUrl)) {
            die('Target url type not valid'.self::NL);
        }

        $this->config['url'] = filter_var($completeUrl, FILTER_SANITIZE_URL);

        if (!$this->valFrequency($this->config['frequency'])) {
            die('Frequency value not valid'.self::NL);
        }
        if (!$this->valPriority($this->config['priority'])) {
            die('Priority value not valid'.self::NL);
        }
        if (!isset($this->config['extension'])) {
            $this->config['extension'] = $this->defaultExtensions;
        }
    }

    private function printConfig($json = false)
    {
        if (!$json) {
            echo '[Crawler Config Variables]'.self::NL;
            foreach ($this->config as $option => $value) {
                $value = is_array($value) ? implode(', ', $value) : $value;
                echo $option.' => '.$value.self::NL;
            }
            echo self::NL;
        } else {
            return json_encode($this->config);
        }
        die();
    }

    public function startCrawl()
    {
        if (!$this->ready) {
            $this->loadConfig();
        }
        $this->createFile();
        $this->doScan($this->config['url']);
        $this->saveAndClose();

        return;
    }

    protected function getUrl($url)
    {
        $agent = 'Mozilla/5.0(compatible;)';
        $curlHandler = curl_init();
        curl_setopt($curlHandler, CURLOPT_AUTOREFERER, true);
        curl_setopt($curlHandler, CURLOPT_URL, $url);
        curl_setopt($curlHandler, CURLOPT_USERAGENT, $agent);
        curl_setopt($curlHandler, CURLOPT_SSL_VERIFYPEER, false);
        curl_setopt($curlHandler, CURLOPT_SSL_VERIFYHOST, false);
        curl_setopt($curlHandler, CURLOPT_HEADER, 0);
        curl_setopt($curlHandler, CURLOPT_FOLLOWLOCATION, 1);
        curl_setopt($curlHandler, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($curlHandler, CURLOPT_CONNECTTIMEOUT, 5);

        $data = curl_exec($curlHandler);

        curl_close($curlHandler);
        return $data;
    }

    protected function clearUrlList(array $list)
    {
        if (empty($list)) {
            return;
        }

        $params = parse_url($this->config['url']);
        $needle = $params['host'];
        $lenght = strlen($this->config['url']);
        if (stripos($params['host'], 'www') !== false) {
            $needle = substr($params['host'], 4);
        }
        unset($params);

        //remove duplicates
        $list = array_merge(array_flip(array_flip($list)));
        
        //transform relative links to absolute, remove links to external websites and to subdomains
        foreach ($list as $i => $link) {
            //rel to abs - basic version
            if (stripos($link, "/") === 0) {
                $link = $this->config['url'].$link;
            }
            //no subdomains or external websites
            if (substr($link, 0, $lenght) !== $this->config['url']) {
                $link = $this->config['url'];
            }
            $list[$i] = filter_var($link, FILTER_SANITIZE_URL);
        }

        //remove all links that have already been scanned.
        foreach ($list as $i => $link) {
            if (in_array($link, $this->scanned)) {
                unset($list[$i]);
            }
        }

        //remove duplicates again
        $list = array_merge(array_flip(array_flip($list)));

        return $list;
    }

    protected function doScan($url)
    {
        if (in_array($url, $this->scanned)) {
            return;
        }
        echo self::NL.'Scanning: '.$url;
        array_push($this->scanned, $url);

        $dom = new \DOMDocument();
        $internalErrors = libxml_use_internal_errors(true);
        $dom->loadHTML($this->getUrl($url));
        libxml_use_internal_errors($internalErrors);
        $links = array();
        foreach ($dom->getElementsByTagName('a') as $link) {
            $links[] = $link->getAttribute('href');
        }

        $links = $this->clearUrlList($links);
        
        if (empty($links)) {
            return;
        }
        foreach ($links as $val) {
            $nextUrl = $val or '';

            foreach ($this->config['extension'] as $ext) {
                if (strpos($nextUrl, trim($ext)) > 0) {
                    fwrite(
                        $this->fileHandle,
                        '
                        <url>
                            <loc>'.htmlentities($nextUrl).'</loc>
                            <changefreq>'.$this->config['frequency'].'</changefreq>
                            <priority>'.$this->config['priority'].'</priority>
                        </url>'
                    );
                    $this->doScan($nextUrl);
                }
            }
        }
    }

    private function createFile()
    {
        $this->fileHandle = fopen($this->config['filename'], 'w');
        if (!$this->fileHandle) {
            die('Sitemap File not available'.self::NL);
        }

        fwrite(
            $this->fileHandle,
            '<?xml version="1.0" encoding="UTF-8"?>
            <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
                    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                    xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9
                    http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd">
            <url>'.self::NL.'
            <loc>'.htmlentities($this->config['url']).'</loc>
                <changefreq>'.$this->config['frequency'].'</changefreq>
                <priority>'.$this->config['priority'].'</priority>
            </url>'.self::NL
        );

        return;
    }

    private function saveAndClose()
    {
        if (!$this->fileHandle) {
            die("Can't file sitemap file to close".self::NL);
        }
        fwrite($this->fileHandle, '</urlset>'.self::NL);
        fclose($this->fileHandle);

        echo self::NL.'Done.'.self::NL.$this->config['filename'].' created.'.self::NL;
    }

    protected function valProtocol($val)
    {
        $protocols = array('http', 'https');

        return in_array($val, $protocols);
    }

    protected function valFilename($val)
    {
        $pattern = '/^(?!.*\/)(\w|\s|-)+\.xml$/';

        return preg_match($pattern, $val) == 1 ? true : false;
    }

    protected function valFrequency($val)
    {
        $frequencies = array('always', 'hourly', 'daily', 'weekly', 'monthly', 'yearly', 'never');

        return in_array($val, $frequencies);
    }

    protected function valPriority($val)
    {
        if (filter_var($val, FILTER_VALIDATE_FLOAT) <= 1 && is_numeric($val)) {
            return $val > 0 ? true : false;
        }

        return false;
    }

    protected function valUrl($url)
    {
        if (!preg_match('%^(?:(?:https?)://)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\x{00a1}-\x{ffff}0-9]-*)*[a-z\x{00a1}-\x{ffff}0-9]+)(?:\.(?:[a-z\x{00a1}-\x{ffff}0-9]-*)*[a-z\x{00a1}-\x{ffff}0-9]+)*(?:\.(?:[a-z\x{00a1}-\x{ffff}]{2,}))\.?)(?::\d{2,5})?(?:[/?#]\S*)?$%uiS', $url)) {
            return false;
        }

        $curlHandler = curl_init($url);
        curl_setopt($curlHandler, CURLOPT_HEADER, true);
        curl_setopt($curlHandler, CURLOPT_NOBODY, true);
        curl_setopt($curlHandler, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($curlHandler, CURLOPT_FOLLOWLOCATION, 1);
        curl_setopt($curlHandler, CURLOPT_TIMEOUT, 10);
        $output = curl_exec($curlHandler);
        $httpCode = curl_getinfo($curlHandler, CURLINFO_HTTP_CODE);
        curl_close($curlHandler);

        if ($httpCode == "200") {
            return true;
        }
        
        return false;
    }


    private function printHelp()
    {
        echo '
Dead Simple Sitemap Generator version '.self::VERSION.'
Giovanni Savastano(gsavastano@gmail.com)
MIT Licence
Use at your own risk :)

Usage: php sitemap.php [options] ...

Option  Meaning
-v  Display program version
-h  Print Help 
-h  Print Crawler Config
-s  Set Output file name  
-a  Set Procol
-u  Set Target URL 
-f  Set Frequency
-p  Set Priority
';
        die();
    }

    private function printVersion()
    {
        echo '
Dead Simple Sitemap Generator version '.self::VERSION.'
Giovanni Savastano(gsavastano@gmail.com)
MIT Licence
Use at your own risk :)
';
        die();
    }
}
